spice/spice.py

```diff
@@ -11,101 +11,128 @@
+    text: str
+    total_time: float
+    input_tokens: int
+    output_tokens: int
+    completed: bool
+    # TODO: Add cost
+
+    @property
+    def total_tokens(self) -> int:
+        return self.input_tokens + self.output_tokens
+
+    @property
+    def characters_per_second(self) -> float:
+        return len(self.text) / self.total_time
 
-        self._stream: Optional[Callable[[], AsyncIterator[str]]] = None
-        self._text = None
+
+class StreamingSpiceResponse:
+    """
+    Returned from a streaming llm call. Can be iterated over asynchronously to retrieve the content.
+    """
+
+    def __init__(self, call_args: SpiceCallArgs, client: WrappedClient, stream: AsyncIterator):
+        self._call_args = call_args
+        self._text = []
         self._start_time = timer()
-        self._first_token_time: Optional[float] = None
         self._end_time = None
         self._input_tokens = None
         self._output_tokens = None
+        self._finished = False
+        self._client = client
+        self._stream = stream
 
-    def finalize(self, text: str, input_tokens: Optional[int], output_tokens: Optional[int]):
-        self._end_time = timer()
-        self._text = text
-        # TODO: this message counting methods are just for OpenAI,
-        # if other providers also don't send token counts when streaming
-        # then we need to add counting methods for them as well.
-        # TODO: input/output will also be none if streaming is interrupted
-        # so we need to identify the provider here to handle correctly
-        if input_tokens is None:
-            self._input_tokens = count_messages_tokens(self.call_args.messages, self.call_args.model)
-        else:
-            self._input_tokens = input_tokens
-        if output_tokens is None:
-            self._output_tokens = count_string_tokens(self.text, self.call_args.model, full_message=False)
-        else:
-            self._output_tokens = output_tokens
-        if self._logging_callback is not None:
-            self._logging_callback(self)
-
-    @property
-    def stream(self) -> Callable[[], AsyncIterator[str]]:
-        if self._stream is None:
-            raise SpiceError("Stream not set! Did you use stream=True?")
-        return self._stream
+    def __aiter__(self):
+        return self
 
-    @property
-    def text(self) -> str:
-        if self._text is None:
-            raise SpiceError("Text not set! Did you iterate over the stream?")
-        return self._text
-
-    @property
-    def time_to_first_token(self) -> float:
-        if self._stream is None or self._first_token_time is None:
-            raise SpiceError("Time to first token not tracked for non-streaming responses")
-        return self._first_token_time - self._start_time
-
-    @property
-    def total_time(self) -> float:
+    async def __anext__(self):
+        # TODO: Do we need to catch errors here? I think it should only be necessary in actual api call
```

Comment:
hmm most of these probably happen immediately but some of them could occur as you stream, like `InternalServerError`. Would be really rare though

<img width="847" alt="image" src="https://github.com/AbanteAI/spice/assets/128252497/646cb164-e6c5-4bcd-b4ce-8fdcf326426d">

Reply:
Hmm true. Might as well keep it then I guess.